File: sol/solution/0001-0100/0039. Combination Sum/Solution.go
func combinationSum(candidates []int, target int) [][]int {
    sort.Ints(candidates)
    var result [][]int
    var current []int
    var backtrack func(start, target int)
    backtrack = func(start, target int) {
        if target == 0 {
            comb := make([]int, len(current))
            copy(comb, current)
            result = append(result, comb)
            return
        }
        for i := start; i < len(candidates); i++ {
            c := candidates[i]
            if c > target {
                break
            }
            current = append(current, c)
            backtrack(i, target-c) 
            current = current[:len(current)-1]
        }
    }
    backtrack(0, target)
    return result
}

File: sol/solution/0001-0100/0039. Combination Sum/Solution.rb
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum(candidates, target)
  candidates.sort!
  result = []
  current = []
  backtrack = lambda do |start, remaining|
    if remaining == 0
      result << current.dup
      return
    end
    (start...candidates.size).each do |i|
      num = candidates[i]
      break if num > remaining
      current << num
      backtrack.call(i, remaining - num)
      current.pop
    end
  end
  backtrack.call(0, target)
  result
end